### DISCUSSION TITLE
CLI flags force interactive mode

### DISCUSSION THREAD

**Discussion by OJFord** [serverless/serverless#11275] [Q&A]
### Are you certain it's a bug?

- [X] Yes, it looks like a bug

### Is the issue caused by a plugin?

- [X] It is not a plugin issue

### Are you using the latest v3 release?

- [X] Yes, I'm using the latest v3 release

### Is there an existing issue for this?

- [X] I have searched existing issues, it hasn't been reported yet

### Issue description

Supplying any CLI flags/arguments forces interactive mode. I've (inadvertently) reproduced with:

- `--verbose`
- `--debug='*'`
- `--use-local-credentials`

Effectively meaning these can't be used in CI.

I haven't tried to use any that _have_ worked, so I assume it's any.

### Service configuration (serverless.yml) content

```yaml
N/A
```

### Command name and used flags

sls --verbose deploy

### Command output

```shell
Attempted to run an interactive setup in non TTY environment.
If that's intended, run with the SLS_INTERACTIVE_SETUP_ENABLE=1 environment variable
error Command failed with exit code 1.
```

### Environment information

```shell
Environment: linux, node 18.5.0, framework 3.15.0 (local), plugin 6.2.2, SDK 4.3.2
```

**Comment by medikoo** [comment]
@OJFord with v3, option names must follow command name and not other way around (it's to avoid ambiguity in options resolution, and it's also what's followed in other popular CLI's).

So, rely on `sls deploy --verbose` instead

**Reply by OJFord** [reply]
Ok, thanks. Strange result for doing it 'wrong' though.

Fwiw, in my experience more common is that 'global' args precede the subcommand (or sometimes can go anywhere), and yes, of course `deploy` specific args follow it.

**Reply by OJFord** [reply]
I'm not really sure I agree this is a support request / discussion by the way - your answer is a fine work-around or fix for why I ran into the problem - but surely the way it presents is a bug.

For example, `sls --anything` could give me an error 'arguments must follow subcommand' or whatever, and it would be much clearer.

**Reply by medikoo** [reply]
> but surely the way it presents is a bug.

It's not a bug, it's intended behavior and I've outlined the reasoning behind that.

> For example, sls --anything could give me an error 'arguments must follow subcommand' or whatever, and it would be much clearer.

There's no way to tell, what was the user intention. e.g. picture `sls --foo bar`, we can't be 100% sure whether intention is to use have `bar` as value for `--foo` param, or whether it's `bar` command with `--foo` boolean param.

To avoid such ambiguity traps, and provide reliable CLI params validation, order is now fixed to `sls [command] [params]`

**Reply by OJFord** [reply]
I understand your reasoning even if I don't agree - what I mean by 'the way it presents' is the fact that it errors about trying to run interactive onboarding.

> To avoid such ambiguity traps, and provide reliable CLI params validation, order is now fixed to sls [command] [params]

Right, so `sls --foo <anything>` is _always_ invalid, so my suggestion is that you can unambiguously error (more appropriately than it does) on `sls --foo` with a help/usage message much like you describe.

**Reply by OJFord** [reply]
When used with serverless-finch for example `sls client --verbose deploy`, this makes it silently fail/succeed without doing anything - essentially running `sls client --verbose` I suppose.

**Reply by medikoo** [reply]
> Right, so sls --foo <anything> is always invalid

That's perfectly valid, if `<anything>` is meant to be a value for `--foo` param, or another param.

What's invalid is `sls [...params] [command]`

> When used with serverless-finch for example

This I believe is a bug in `serverless-finch` plugin, which is best if it's reported there.

**Reply by OJFord** [reply]
Ok I think I see - you are saying they are valid before a subcommand if they take an argument, and it is provided? Otherwise, the subcommand is parsed as the argument, so it's not valid and argumentless flag needs to follow subcommand?

Why then is this reproducible with `--debug='*'`?

Look all that aside, my basic point is that it shouldn't error about interactive onboarding. It should show usage and exit 1 or whatever. I gather it's effectively running bare `sls`, and as a result trying to ask 'what do you want to do'. Presently it's confusing and misleading. I spent hours on two approaches: 1) hackily answering interactive questions non-interactively; 2) trying in vain to disable interactive mode. I didn't expect _removing_ debug/verbose logging would help - there was no indication there was any argument parsing error, or missing subcommand (parsed as arg value) - even when interactive enabled (via env var) and proceeded with that.

**Comment by medikoo** [comment]
> Ok I think I see - you are saying they are valid before a subcommand if they take an argument, a

They're never valid before command

**Reply by OJFord** [reply]
That's not what you said above when I said that.

I have no interest in going around on circles on this, I'm just trying to offer some feedback to make it more usable.
